sdcm/chaos_mesh.py

```diff
@@ -0,0 +1,62 @@
+import json
+import logging
+
+from sdcm.cluster_k8s import KubernetesCluster
+from sdcm.utils.k8s import HelmValues, KubernetesOps
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ChaosMesh:  # pylint: disable=too-few-public-methods
+    NAMESPACE = "chaos-mesh"
+    VERSION = "2.5.0"
+    HELM_SETTINGS = {'dashboard.create': True, 'dnsServer.create': True}
+
+    def __init__(self):
+        self._k8s_cluster = None
+
+    def initialize(self, k8s_cluster: KubernetesCluster) -> None:
+        """Installs chaos-mesh on k8s cluster and prepares for future k8s chaos testing."""
+        if self._k8s_cluster is not None:
+            LOGGER.warning("Chaos Mesh is already initialized. It should have not been initialized twice.")
+            return
+        self._k8s_cluster = k8s_cluster
+        self._install()
+
+    def _install(self):
+        if self._k8s_cluster.kubectl("kubectl get ns chaos-mesh", ignore_status=True).ok:
+            LOGGER.info("Chaos Mesh is already installed. Skipping installation.")
+            return
+        LOGGER.info("Installing chaos-mesh...")
+        self._k8s_cluster.helm("repo add chaos-mesh https://charts.chaos-mesh.org")
+        self._k8s_cluster.helm('repo update')
+        self._k8s_cluster.kubectl("create namespace chaos-mesh")
+        self._k8s_cluster.helm_install(
+            target_chart_name="chaos-mesh",
+            source_chart_name="chaos-mesh/chaos-mesh",
+            version=self.VERSION,
+            use_devel=False,
+            namespace=self.NAMESPACE,
+            values=HelmValues(self.HELM_SETTINGS),
+            atomic=True,
+            timeout="20m"
+        )
+        self._wait_for_chaos_mesh_readiness(timeout=20)
+        LOGGER.info("chaos-mesh installed successfully.")
+
+    def _wait_for_chaos_mesh_readiness(self, timeout: int):
+        """Additional wait to make sure all chaos-mesh pods are up and running"""
+        LOGGER.info("waiting for chaos-mesh pods running...")
+        pods_info = json.loads(self._k8s_cluster.kubectl("get pods -o json", namespace="chaos-mesh").stdout)
+        total_pods = len(pods_info["items"])
+        KubernetesOps.wait_for_pods_readiness(
+            kluster=self._k8s_cluster,
+            total_pods=total_pods,
+            readiness_timeout=timeout,
+            namespace="chaos-mesh"
+        )
+        LOGGER.info("chaos-mesh pods are running.")
+
+
+# used as singleton - import 'chaos_mesh' instead of ChaosMesh
+chaos_mesh = ChaosMesh()
```

Comment:
I was thinking about it, but decided to do it in setUp directly as I'd need to implement it in all `get_cluster_k8s...` functions - while it should not depend on any k8s backend. Also it is more related to testing than k8s cluster and hiding it somewhere in deployment for me was not looking good.

Btw. Is this possible for scylla to communicate across multiple clusters? Even if we support multiregion, won't it be in one k8s cluster spreading across regions?

Btw. this is a very lazy way of creating singleton and in case of multiple clusters can be refactored easily.

Reply:
It does depend on the K8S clusters. The helm chart installs it directly to a K8S cluster.
At some milestone of serverless we will have multi-region support which assumes having multiple K8S clusters.

I would expect to have chaos-mesh agent be in each K8S cluster.

So, in summary, having direct relation to each separate K8S cluster we can already make it be part of the appropriate class.

Reply:
I'm reluctant to grow k8s cluster classes. I'd rather keep things as pluggable modules. 
In case of multiple clusters, we could easily modify `ChaosMesh` class to install chaos-mesh on all provided k8s clusters and remember on which one was installed.

Later when running experiments, we could just provide cluster on which we want to run it as argument.

With this approach we may stop modifying our classes indefinitely and focus of adding new ones with single responsibilities - which is more maintainable way.

Reply:
Saying be part of a class I meant following:
```
from sdcm.utils.chaos_mesh import ChaosMesh

class KubernetesCluster():
    def __init__(self):
        self.chaos_mesh = ChaosMesh()
```

Reply:
> Saying be part of a class I meant following:
> 
> ```
> from sdcm.utils.chaos_mesh import ChaosMesh
> 
> class KubernetesCluster():
>     def __init__(self):
>         self.chaos_mesh = ChaosMesh()
> ```

Yes, It looks easy, but it breaks good programming principles:
1. Single responsibility principle - KubernetesCluster class will be responsible for chaos-mesh
2. It creates dependencies inside own code - tight coupling
3.  Later, when using it, we'll do `self.cluster.k8s_cluster.chaos_mesh.run_experiment(...)` which also breaks some good design principle (Demeter law)

I'm opting for such approach:
`chaos_mesh.run_experiment(self.cluster.k8s_cluster, experiment)`

Reply:
Still, because it's installed in k8s cluster doesn't mean it must be part of `KubernetesCluster` class - we can inject instance of it to chaos-mesh and have loose coupling.
I just try to avoid the mess like we have in `Cluster` class where we do everything. Easy implementation is not equal easy maintenance (over time rather opposite).